that_depends/providers/base.py

```diff
@@ -214,23 +204,70 @@ def sync_resolve(self) -> T_co:
-            context.context_stack = contextlib.ExitStack()
-            context.instance = context.context_stack.enter_context(
-                contextlib.contextmanager(self._creator)(
-                    *[  # type: ignore[arg-type]
-                        x.sync_resolve() if isinstance(x, AbstractProvider) else x for x in self._args
-                    ],
-                    **{  # type: ignore[arg-type]
-                        k: v.sync_resolve() if isinstance(v, AbstractProvider) else v for k, v in self._kwargs.items()
-                    },
-                ),
-            )
-        return typing.cast(T_co, context.instance)
+        context.context_stack = contextlib.ExitStack()
+        context.instance = context.context_stack.enter_context(cm)
+
+        return context.instance
+
+
+# TODO(zerlok): simplify code in function # noqa: TD003, FIX002
+def _normalize_creator(  # noqa: C901
+    resource: object,
+    creator: ResourceCreator[P, T_co],
+    *args: P.args,
+    **kwargs: P.kwargs,
+) -> tuple[bool, typing.Callable[P, typing.ContextManager[T_co] | typing.AsyncContextManager[T_co]]]:
+    if isinstance(creator, typing.AsyncContextManager):
+        if args or kwargs:
+            msg = "AsyncContextManager does not accept any arguments"
+            raise TypeError(msg, creator, args, kwargs)
+
+        def create_async(*_: P.args, **__: P.kwargs) -> typing.AsyncContextManager[T_co]:
+            return creator
+
+        return True, create_async
+
+    if isinstance(creator, typing.ContextManager):
+        if args or kwargs:
+            msg = "ContextManager does not accept any arguments"
+            raise TypeError(msg, creator, args, kwargs)
+
+        def create_sync(*_: P.args, **__: P.kwargs) -> typing.ContextManager[T_co]:
+            return creator
+
+        return False, create_sync
+
+    if inspect.isasyncgenfunction(creator):
+        return True, contextlib.asynccontextmanager(creator)
+
+    if inspect.isgeneratorfunction(creator):
+        return False, contextlib.contextmanager(creator)
+
+    if inspect.isfunction(creator):
+        returns = inspect.signature(creator).return_annotation
+        returns_origin = typing.get_origin(returns) or returns
+
+        # NOTE: creator may be a wrapped with `asynccontextmanager` or `contextmanager` decorator, but it's not possible
+        # to reveal it here. Assume that mypy covered that cases.
+        if issubclass(returns_origin, typing.AsyncContextManager | typing.AsyncIterator):
+            return True, typing.cast(typing.Callable[P, typing.AsyncContextManager[T_co]], creator)
+
+        if issubclass(returns_origin, typing.ContextManager | typing.Iterator):
+            return False, typing.cast(typing.Callable[P, typing.ContextManager[T_co]], creator)
+
+    # TODO(zerlok): suggest use TypeError instead, backward incompatible change. # noqa: TD003, FIX002
+    msg = f"{type(resource).__name__} must be generator function"
+    raise RuntimeError(msg)
```

Comment:
better use TypeError, because `creator` is not of valid type here

Reply:
I agree. Let's make it TypeError